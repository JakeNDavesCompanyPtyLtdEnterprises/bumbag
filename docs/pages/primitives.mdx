import { Box, Block, Inline, InlineBlock, InlineFlex, Grid, Flex } from 'fannypack';

# Primitives

Fannypack exports Reakit's primitives:

- `<Box>`
- `<Block>`
- `<Inline>`
- `<InlineBlock>`
- `<Flex>`
- `<InlineFlex>`
- `<Grid>`

These primitive components provide an abstraction on top of the `<div>` element to include reset styles and allows you to specify CSS attributes as props (see the below for more details).

See more about primitive components [here](https://reakit.io/components/box).

## Box

```jsx
import { Box } from 'fannypack'
```

All components in Fannypack are built on top of the `<Box>` primitive. Any component in Fannypack can use `Box`'s props. [Read more here](https://reakit.io/components/box).

By default it renders a `<div>` element with basic reset styles.

```.jsx
<Box>This is a box</Box>
```

You're able to apply any CSS attribute as a prop to box:

```.jsx
<Box backgroundColor="primary" color="white" padding="major-3">
  This is a box
</Box>
```

#### Color attributes

Any CSS color attribute (e.g. `color`, `border-color`, etc) can take a [color from the Fannypack palette](/palette#theming).

```.jsx
<Box backgroundColor="primaryTint" color="primary">
  This is a box
</Box>
```

#### Margin & padding attributes

Any CSS margin/padding attribute (e.g. `margin`, `padding-left`, etc) can take a [spacing from the Fannypack theme](/styling#layoutthemeconfig).

```.jsx
<Box backgroundColor="primaryTint" padding="major-3">
  This is a box
</Box>
```

#### Font attributes

The `fontWeight` CSS prop can contain of a [font weight in the Fannypack theme](/styling#fontweightsthemeconfig).

```.jsx
<Box fontWeight="semibold">This is a box</Box>
```

The `fontSize` CSS prop can contain of a [font size in the Fannypack theme](/styling#fontsizethemeconfig).

```.jsx
<Box fontSize="400">This is a box</Box>
```

## Block

```jsx
import { Block } from 'fannypack'
```

A `<Block>` is a `<Box>` with `display: block`. [Read more here](https://reakit.io/components/block).

```.jsx
<Block backgroundColor="primaryTint" padding="major-3">
  Block 1
</Block>
<Block backgroundColor="successTint" padding="major-3">
  Block 2
</Block>
<Block backgroundColor="dangerTint" padding="major-3">
  Block 3
</Block>
```

## Inline

```jsx
import { Inline } from 'fannypack'
```

An `<Inline>` is a `<Box>` with `display: inline`. [Read more here](https://reakit.io/components/inline).

```.jsx
<Inline>
  Inline 1
</Inline>
<Inline>
  Inline 2
</Inline>
<Inline>
  Inline 3
</Inline>
```

## InlineBlock

```jsx
import { InlineBlock } from 'fannypack'
```

An `<InlineBlock>` is a `<Box>` with `display: inline-block`. [Read more here](https://reakit.io/components/inlineblock).

```.jsx
<InlineBlock backgroundColor="primaryTint" padding="major-3">
  InlineBlock 1
</InlineBlock>
<InlineBlock backgroundColor="successTint" padding="major-3">
  InlineBlock 2
</InlineBlock>
<InlineBlock backgroundColor="dangerTint" padding="major-3">
  InlineBlock 3
</InlineBlock>
```

## Flex

```jsx
import { Flex } from 'fannypack'
```

A `<Flex>` is a `<Box>` with `display: flex`. [Read more here](https://reakit.io/components/flex).

```.jsx
<Flex>
  <Box backgroundColor="primaryTint" padding="major-3">
    Box 1
  </Box>
  <Box backgroundColor="successTint" padding="major-3">
    Box 2
  </Box>
  <Box backgroundColor="dangerTint" padding="major-3">
    Box 3
  </Box>
</Flex>
```

## InlineFlex

```jsx
import { InlineFlex } from 'fannypack'
```

An `<InlineFlex>` is a `<Box>` with `display: inline-flex`. [Read more here](https://reakit.io/components/inlineflex).

```.jsx
<InlineFlex justifyContent="space-evenly" width="100%">
  <Box backgroundColor="primaryTint" padding="major-3">
    Box 1
  </Box>
  <Box backgroundColor="successTint" padding="major-3">
    Box 2
  </Box>
  <Box backgroundColor="dangerTint" padding="major-3">
    Box 3
  </Box>
</InlineFlex>
```

## Grid

```jsx
import { Grid } from 'fannypack'
```

A `<Grid>` is a `<Box>` with `display: grid`. It also consists of a `<Grid.Item>` component to pair with the `<Grid>`. [Read more here](https://reakit.io/components/grid).

```.jsx
<Grid templateColumns="repeat(2, 1fr)" autoRows="auto" gap="3vw">
  <Grid.Item backgroundColor="primaryTint" padding="major-3">
    Grid Item 1
  </Grid.Item>
  <Grid.Item backgroundColor="successTint" padding="major-3">
    Grid Item 2
  </Grid.Item>
  <Grid.Item backgroundColor="dangerTint" padding="major-3">
    Grid Item 3
  </Grid.Item>
</Grid>
```
