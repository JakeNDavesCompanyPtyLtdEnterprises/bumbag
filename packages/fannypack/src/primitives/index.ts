export { default as Box } from 'reakit/Box';
export { default as Block } from 'reakit/Block';
export { default as Inline } from 'reakit/Inline';
export { default as InlineBlock } from 'reakit/InlineBlock';
export { default as Flex } from 'reakit/Flex';
export { default as InlineFlex } from 'reakit/InlineFlex';
export { default as Grid } from 'reakit/Grid';
