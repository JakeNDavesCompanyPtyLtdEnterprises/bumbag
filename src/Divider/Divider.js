// @flow
import React, { type Node } from 'react';
import _Divider from './styled';

type Props = {
  children: Node,
  className?: string,
  content?: string,
  isVertical?: boolean
};

const Divider = ({ children, className, content, isVertical, ...props }: Props) => (
  <_Divider className={className} content={content} vertical={isVertical} {...props}>
    {children}
  </_Divider>
);

Divider.defaultProps = {
  className: undefined,
  content: undefined,
  isVertical: false
};

export default Divider;
