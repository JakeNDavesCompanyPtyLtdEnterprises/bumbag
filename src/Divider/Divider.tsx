import * as React from 'react';
// @ts-ignore
import PropTypes from 'prop-types';
import { DividerProps as ReakitDividerProps } from 'reakit/ts';

import _Divider from './styled';

export interface LocalDividerProps {
  children?: React.ReactNode;
  className?: string;
  content?: string;
  isVertical?: boolean;
}
export type DividerProps = LocalDividerProps & ReakitDividerProps;

export const Divider: React.FunctionComponent<LocalDividerProps> = ({
  children,
  className,
  content,
  isVertical,
  ...props
}) => (
  <_Divider className={className} content={content} vertical={isVertical} {...props}>
    {children}
  </_Divider>
);

Divider.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  content: PropTypes.string,
  isVertical: PropTypes.bool
};
Divider.defaultProps = {
  children: null,
  className: undefined,
  content: undefined,
  isVertical: false
};

const C: React.FunctionComponent<DividerProps> = Divider;
export default C;
