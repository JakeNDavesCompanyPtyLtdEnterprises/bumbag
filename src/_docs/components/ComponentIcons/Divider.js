import React from 'react';

const SvgComponent = props => (
  <svg width={83} height={1} fill="none" {...props}>
    <path stroke="#000" d="M0 .5h83" />
  </svg>
);

export default SvgComponent;
