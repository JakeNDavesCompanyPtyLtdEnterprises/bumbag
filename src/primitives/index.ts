// @ts-ignore
export { default as Box } from 'reakit/Box';
// @ts-ignore
export { default as Block } from 'reakit/Block';
// @ts-ignore
export { default as Inline } from 'reakit/Inline';
// @ts-ignore
export { default as InlineBlock } from 'reakit/InlineBlock';
// @ts-ignore
export { default as Flex } from 'reakit/Flex';
// @ts-ignore
export { default as InlineFlex } from 'reakit/InlineFlex';
// @ts-ignore
export { default as Grid } from 'reakit/Grid';
