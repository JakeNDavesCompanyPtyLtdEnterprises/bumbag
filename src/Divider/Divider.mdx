---
name: Divider
route: /components/divider
menu: Components
---

import { Playground, PropsTable } from 'docz';
import {Divider} from './Divider';
import { Box, Block, Flex } from '../index';

# Divider

## Import

`import { Divider } from 'fannypack'`

## Basic Usage

`<Divider>` renders a `<hr>` element with reset styles and basic divider styling.

<Playground>
  <Divider />
</Playground>

## Vertical

Render a vertical divider by adding the `isVertical` prop to the component.

<Playground>
  <Flex>
    Hello
    <Divider isVertical />
    World
  </Flex>
</Playground>

## With content

A divider can also have content. The `content` prop accepts a string.

<Playground>
  <Flex alignItems="center" height="50px">
    Hello
    <Divider content="or" isVertical />
    World
  </Flex>
  <Box textAlign="center" marginTop="xsmall">
    Hello
    <Divider content="or" />
    World
  </Box>
</Playground>

## Props

<PropsTable of={Divider} />

## Theming

### Schema

```jsx
{
  base: string | Object,
  content: string | Object,
  vertical: string | Object
}
```
