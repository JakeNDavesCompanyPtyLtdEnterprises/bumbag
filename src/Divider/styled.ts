import { Divider } from 'reakit';
import { theme } from 'styled-tools';

import { styled, css } from '../styled';
import { DividerProps } from './Divider';

const contentProperties = css`
  display: flex;
  align-items: center;
  justify-content: center;

  &::before {
    background: white;
    color: #c3c3c3;
    content: '${(props: any) => props.content}';
    text-transform: uppercase;
    font-size: 0.7em;
    font-weight: ${theme('fannypack.fontWeights.bold')};
    padding: 0.2rem 0.5rem;

    & {
      ${theme('fannypack.Divider.content')}
    }
  }
`;
const verticalProperties = css`
  height: auto;

  & {
    ${theme('fannypack.Divider.vertical')};
  }
`;

export default styled(Divider)<DividerProps>`
  opacity: 1;
  border-color: #e7e9ed;

  & {
    ${props => props.content && contentProperties};
    ${props => props.vertical && verticalProperties};
  }

  & {
    ${theme('fannypack.Divider.base')};
  }
`;
