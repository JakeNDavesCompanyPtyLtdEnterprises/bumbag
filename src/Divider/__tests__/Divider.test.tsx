import * as React from 'react';
import render from '../../_utils/tests/render';
import Divider from '../Divider';
import { Box, Flex } from '../../primitives';

it('renders correctly for a basic divider', () => {
  const { container } = render(<Divider />);
  expect(container.firstChild).toMatchSnapshot();
});

it('renders correctly for a vertical divider', () => {
  const { container } = render(
    <Flex>
      Hello
      <Divider isVertical />
      World
    </Flex>
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders correctly for a horizontal divider with content', () => {
  const { container } = render(
    <Box>
      Hello
      <Divider content="Test" />
      World
    </Box>
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('renders correctly for a vertical divider with content', () => {
  const { container } = render(
    <Flex alignItems="center" height="50px">
      Hello
      <Divider content="or" isVertical />
      World
    </Flex>
  );
  expect(container.firstChild).toMatchSnapshot();
});
