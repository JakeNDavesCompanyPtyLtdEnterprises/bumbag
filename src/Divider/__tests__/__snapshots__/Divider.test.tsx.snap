// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly for a basic divider 1`] = `
.c2 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c2:focus:not(:focus-visible) {
  outline: none;
}

.c1 {
  border-color: currentColor;
  border-style: solid;
  opacity: 0.2;
  margin: 1em 0;
  height: 0;
  border-width: 1px 0 0 0;
}

.c0 {
  opacity: 1;
  border-color: #e7e9ed;
}

<hr
  class="c0 c1 c2"
/>
`;

exports[`renders correctly for a horizontal divider with content 1`] = `
.c0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c0:focus:not(:focus-visible) {
  outline: none;
}

.c2 {
  border-color: currentColor;
  border-style: solid;
  opacity: 0.2;
  margin: 1em 0;
  height: 0;
  border-width: 1px 0 0 0;
}

.c1 {
  opacity: 1;
  border-color: #e7e9ed;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1::before {
  background: white;
  color: #c3c3c3;
  content: 'Test';
  text-transform: uppercase;
  font-size: 0.7em;
  font-weight: 700;
  padding: 0.2rem 0.5rem;
}

<div
  class="c0"
>
  Hello
  <hr
    class="c1 c2 c0"
    content="Test"
    style="content: Test;"
  />
  World
</div>
`;

exports[`renders correctly for a vertical divider 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c3 {
  border-color: currentColor;
  border-style: solid;
  opacity: 0.2;
  margin: 0 1em;
  min-height: 100%;
  width: 0;
  border-width: 0 0 0 1px;
}

.c2 {
  opacity: 1;
  border-color: #e7e9ed;
}

.c2 {
  height: auto;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  class="c0 c1"
>
  Hello
  <hr
    class="c2 c3 c1"
  />
  World
</div>
`;

exports[`renders correctly for a vertical divider with content 1`] = `
.c1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
  background-color: unset;
  color: inherit;
}

.c1:focus:not(:focus-visible) {
  outline: none;
}

.c3 {
  border-color: currentColor;
  border-style: solid;
  opacity: 0.2;
  margin: 0 1em;
  min-height: 100%;
  width: 0;
  border-width: 0 0 0 1px;
}

.c2 {
  opacity: 1;
  border-color: #e7e9ed;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: auto;
}

.c2::before {
  background: white;
  color: #c3c3c3;
  content: 'or';
  text-transform: uppercase;
  font-size: 0.7em;
  font-weight: 700;
  padding: 0.2rem 0.5rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  class="c0 c1"
  height="50px"
  style="align-items: center; height: 50px;"
>
  Hello
  <hr
    class="c2 c3 c1"
    content="or"
    style="content: or;"
  />
  World
</div>
`;
